### ISSUE TITLE
Basic Recursion: Simpler solution

### ISSUE THREAD

**Issue by 17xande** [timoxley/functional-javascript-workshop#32]
I may have found a simpler solution to the recursion problem:

``` JavaScript
function reduce(arr, fn, initial) {
    if (arr.length === 0) {
        return initial;
    }

    return fn(initial, arr.shift(), 0, arr);
}

module.exports = reduce;
```

I'm not sure how this solution compares to the one provided, perhaps it's less efficient, or maybe it doesn't work on all cases, but it makes more sense in my head, mostly because it seems a bit simpler. It also passes the test successfully.

Additionally, I had to read the problem description like 10 times before I thought I understood what was required.
i.e. Part of the description reads: "... and an _initial value which will return an object_ containing the counts for each word found in the array ...".
That sounds like the initial value will return an object? So my first thought was: _is the initial value a function as well?_
Eventually I figured it out. Maybe I'm the only one that struggled with the description, otherwise, perhaps we could improve it slightly? Just a thought...

**Comment by kierans** [comment]
My first attempt at the Basic Recursion exercise was very similar to @Monocotyledon however I defaulted the index value to _fn_ to zero initially.  Running my code through the verification process, my program passed.  Eventually I thought of how to increment the index properly; for the sake of completing the problem properly.  However the verification step should fail if the index isn't correct for each invocation of _fn_ IMO since you're not meeting the requirements.

I'm guessing based on this thread that the calculation of the index is the hardest part of the problem, so feedback for users would also be useful.

I also think the inlining of the function naming/using in the official solution is harder to read, but that might be a personal preference.  Here's my solution:

    function reduce(arr, fn, initial) {
      function reduceWithCounter(counter, value) {
        if (counter > arr.length - 1) {
          return value;
        }
    
        return reduceWithCounter(counter + 1, fn(value, arr[counter], counter, arr));
      }
    
      return reduceWithCounter(0, initial);
    }

**Comment by alex7217** [comment]
Hopping onto this issue with a similar feedback to what @kierans and @Monocotyledon said.
1. Awesome workshop so far - really enjoy the format and challenges.
2. When I arrived at  Basic: Recursion, Exercise 7 of 18 - the directions kinda baffled me. I did not understand the part:

> You don't need to implement this functionality, it will be supplied to your reduce implementation.

How or why the code I wrote from the previous exercise would be used exactly was mysterious, and compounded my difficulties to grasp recursion concepts.

3. Finally, like @Monocotyledon said, the official solution I don't think I would have ever arrived at. A hint about IIFE or maybe a previous exercise involving them would help.

TL;DR - The conceptual leap between exercises 6 and 7 may be too great for JavaScript beginners. We need more direct instructions or additional clues about the solution.

**Comment by RobBollons** [comment]
Just adding to the feedback about this issue. I've enjoyed the exercises so far but the official solution for this one looks a bit messy to me and isn't easy to follow due to the IIFE and the inlining. In the end I got the same solution as @Monocotyledon which I guess by using Array.prototype.slice isn't optimal. I would definitely have needed more pointers about maintaining the index in the recursion like @kierans said.

**Comment by LordJohn42** [comment]
Attention, spoiler. Official solution. 
```
function reduce(arr, fn, initial) {
  return (function reduceOne(index, value) {
    if (index > arr.length - 1) return value // end condition
    return reduceOne(index + 1, fn(value, arr[index], index, arr)) // calculate & pass values to next step
  })(0, initial) // IIFE. kick off recursion with initial values
}
```

**Comment by sovkay** [comment]
> @timoxley ah, true. We can increment the index, albeit at the expense of concision:
> 
> ```js
>   // reduce `arr` with `reduction` method
>   // initialize with `init`
>   var reduce = function(arr, reduction, init) {
>     var i, next, reduced;
>     if (!arr.length) return init                        // nothing left to recur
>     if (typeof i === "undefined" || i === null) i = 0   // increment index
>     next = arr.shift();                                 // shift off first val
>     reduced = reduction(init, next, i + 1, arr);        // get reduced value
>     return reduce(arr, reduction, reduced);             // recur
>   };
> ```

why are you passing arr in the last recursive step [return reduce...] and the function still working. If you do this your base condition shouldn't work.Right?
